Add support for transformer layout of masked_softmax

zrphercule opened the pull request:
Summary:
In transformer encoder and MHA, masked_softmax's mask is a 2D tensor (B, D), where input is a 4D tensor (B, H, D, D).
This mask could be simply broadcasted to a (B, H, D, D) like input, and then do a regular masked_softmax, however it will bring the problem of non-contiguous mask & consume more memory.
In this diff, we maintained mask's shape unchanged, while calc the corresponding mask for input in each cuda thread.

This new layout is not currently supported in CPU yet.

Test Plan: buck build mode/opt -c fbcode.enable_gpu_sections=true caffe2/test:nn && buck-out/gen/caffe2/test/nn\#binary.par -r test_masked_softmax

Differential Revision: D32605557

vadimkantorov commented:
I think we discussed somewhere also general mask expansion / broadcasting, it would be nice to support it

zrphercule commented:
> I think we discussed somewhere also general mask expansion / broadcasting, it would be nice to support it

This diff is specialized for Transformer encoder's need for masked_softmax (input BHDD, mask BD)
Although a general mask broadcasting can be used for more scenarios, it inevitably consume more memory, also has worse performance. So far we encourage user to broadcast mask before calling masked_softmax, however we may come back to this adding a general broadcasting solution as well :)

ngimel commented:
This looks good, but there are a lot of test failures (likely not related to this PR), can you try rebasing?

zrphercule2 committed:
Add support for transformer layout of masked_softmax (#69272)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/69272

In transformer encoder and MHA, masked_softmax's mask is a 2D tensor (B, D), where input is a 4D tensor (B, H, D, D).
This mask could be simply broadcasted to a (B, H, D, D) like input, and then do a regular masked_softmax, however it will bring the problem of non-contiguous mask & consume more memory.
In this diff, we maintained mask's shape unchanged, while calc the corresponding mask for input in each cuda thread.

This new layout is not currently supported in CPU yet.

Test Plan: buck build mode/opt -c fbcode.enable_gpu_sections=true caffe2/test:nn && buck-out/gen/caffe2/test/nn\#binary.par -r test_masked_softmax

Reviewed By: ngimel

Differential Revision: D32605557

fbshipit-source-id: c927e1942dde5dbfa8ba1d01e94148b4255a3f15
